micrometer-observation/src/main/java/io/micrometer/observation/Observation.java

```diff
@@ -120,21 +124,24 @@ static Observation createNotStarted(String name, @Nullable Context context,
      * picked.
      * @param defaultConvention default convention when no custom convention was passed,
      * nor a configured one was found
-     * @param context the observation context
+     * @param contextSupplier supplier for the observation context
      * @param registry observation registry
      * @return created but not started observation
      */
     static <T extends Context> Observation createNotStarted(@Nullable ObservationConvention<T> customConvention,
-            @NonNull ObservationConvention<T> defaultConvention, @NonNull T context,
-            @NonNull ObservationRegistry registry) {
+            ObservationConvention<T> defaultConvention, Supplier<T> contextSupplier, ObservationRegistry registry) {
+        if (registry.isNoop()) {
+            return Observation.NOOP;
+        }
         ObservationConvention<T> convention;
         if (customConvention != null) {
             convention = customConvention;
         }
         else {
-            convention = registry.observationConfig().getObservationConvention(context, defaultConvention);
+            convention = registry.observationConfig().getObservationConvention(contextSupplier.get(),
```

Comment:
This will result in the supplier being called twice, which feels less than ideal.

Reply:
Can we review the noop checks for the whole class, probably refactor it into  a method (or two) and do the registry based check early?
After we passed the registry based check, we can create the context once and pass it to the next method (a private overload without the supplier)?

Reply:
One issue is interfaces can't have private methods. But regardless of that, it's a little hard to centralize this much more than it is, I think, due to the variety of factory methods and the different logic they have. Suggestions welcome, though.

Reply:
What if we create a dedicated factory/builder class instead of default/static factory methods? The class would centralize noop checks for all variants of observation creations if possible.

Reply:
Looking at this more, we have two constructors for SimpleObservation, so I think we're going to have at least two places in factory methods we need to check for a no-op registry. It turns out we end up with 3 now because there is the option to make an Observation with a name, with a convention, or to have the convention looked up. The convention look-up requires the context, so we need to get the context already at that point. I've refactored this method to not call one of the other methods and instead instantiate the SimpleObservation itself, which avoids supplying the context twice. It leads to a little code duplication, but the look-up of the convention in the registry is a notable difference that happens in the middle of the logic.

On a dedicated factory/builder class, that's an option. I worry it's a bigger change and we're basically out of time to iterate on and get feedback on such things. I don't know that it would change a lot. The no-op check that can be centralized I think is just calling `registry.isNoop()` and separately `!registry.observationConfig().isObservationEnabled(name, context)`. This is already only done in the 3 places I mentioned above because they are end up doing different things.

A tangential thing is, we're not consistent on whether we accept a `@Nullable ObservationRegistry` or not in the factory methods. We can look at that separately, I think, but we should make that consistent unless there's some reason to be different.